### PR TITLE
Moved slack client code to mojito-common and created NotificationService

### PR DESCRIPTION
**Pull request by maallen**
Created NotificationService in mojito-common that can be utilised to send notifications to either Phabricator (in the form of a revision comment) or Slack (as an instant message).

Updated PhabricatorExtractionCheckCommand to use the new service for sending failure notifications.

### CODE REVIEW

File: webapp/package-lock.json [hunk 1]
```diff
@@ -6047,9 +6047,12 @@
       },
       "dependencies": {
         "node-fetch": {
-          "version": "2.6.1",
-          "resolved": "https://registry.npmjs.org/node-fetch/-/node-fetch-2.6.1.tgz",
-          "integrity": "sha512-V4aYg89jEoVRxRb2fJdAg8FHvI7cEyYdVAh94HH0UIK8oJxUfkjlDQN9RbMx+bEjP7+ggMiFRprSti032Oipxw=="
+          "version": "2.6.6",
```
Comment (aurambaj) [inline webapp/package-lock.json:6050]: knit: remove or put in an independent commit

**Review by aurambaj — COMMENTED** [review]
first quick pass, I'll review more later

File: common/src/main/java/com/box/l10n/mojito/notifications/service/ThirdPartyService.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package com.box.l10n.mojito.notifications.service;
+
+public enum ThirdPartyService {
```
[the remark this answers is not in the document]
Comment (maallen) [reply common/src/main/java/com/box/l10n/mojito/notifications/service/ThirdPartyService.java:3]: Spring creates the Phabricator and Slack notification services by injecting in relevant configuration variables etc so I don't think we can have the enum create the class instances as we need the config values at runtime. I'll remove the switch statement in `NotificationService` and update it to filter on a list of ThirdPartyNotificationService (injected by Spring) objects instead

File: webapp/package-lock.json (hunk 1 above)
Comment (maallen) [reply webapp/package-lock.json:6050]: This was a mistake, didn't intend to commit it. Will remove

File: common/src/main/java/com/box/l10n/mojito/notifications/service/slack/SlackClient.java [hunk 3]
```diff
@@ -1,11 +1,11 @@
-package com.box.l10n.mojito.slack;
-
-import com.box.l10n.mojito.slack.request.Channel;
-import com.box.l10n.mojito.slack.request.Message;
-import com.box.l10n.mojito.slack.request.User;
-import com.box.l10n.mojito.slack.response.ChatPostMessageResponse;
-import com.box.l10n.mojito.slack.response.ImOpenResponse;
-import com.box.l10n.mojito.slack.response.UserResponse;
+package com.box.l10n.mojito.notifications.service.slack;
```
[the remark this answers is not in the document]
Comment (maallen) [reply common/src/main/java/com/box/l10n/mojito/notifications/service/slack/SlackClient.java]: Added separate commit for moving slack client to mojito-common

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PhabricatorExtractionCheckCommand.java [hunk 4]
```diff
@@ -85,9 +99,12 @@
     @Parameter(names = {"--objectid", "-oid"}, arity = 1, required = false, description = "Phabricator object id, required if adding a comment with check failures to Phabricator.")
```
[the remark this answers is not in the document]
Comment (maallen) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/PhabricatorExtractionCheckCommand.java:99]: The skip checks in emergency option that reviews the phabricator test plan string is still present in the command, I can move that out into it's own separate command that sets an env variable. If the env variable is set we skip the check in our build bash script?

[the remark this answers is not in the document]
Comment (aurambaj) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/PhabricatorExtractionCheckCommand.java:99]: can we add it here `PhabricatorDiffInfoCommand` ?

[the remark this answers is not in the document]
Comment (maallen) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/PhabricatorExtractionCheckCommand.java:99]: It should be possible to add it there and set an env value

File: common/src/main/java/com/box/l10n/mojito/notifications/service/NotificationService.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+public class NotificationService {
+
+    public static final String BOLD_MARKER = "**";
+
+    public static final String ITALIC_MARKER = "--";
+
+    static Logger logger = LoggerFactory.getLogger(NotificationService.class);
+
+    @Autowired(required = false)
+    List<ThirdPartyNotificationService> notificationServices;
```
[the remark this answers is not in the document]
Comment (maallen) [reply common/src/main/java/com/box/l10n/mojito/notifications/service/NotificationService.java:32]: Notification services are configured via spring variables, so env from the command line or via properties file for the server. 

For the PhabricatorExtractionCheckCommand specifically we need to pass in the phabricator diff & slack username to notify as cli parameters as these will differ on each execution. I agree with your later comment about the phabricator parameter on the slack notification params should be moved into the message template itself. I'll need to make two separate calls to the notification service object for each individual notification type as the message notified to slack will be slightly different than that sent to Phabricator

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PhabricatorExtractionCheckCommand.java [hunk 6]
```diff
@@ -151,18 +172,59 @@ protected boolean areChecksSkipped() {
...
+            } catch (NotificationServiceException e) {
+                throw new CommandException("Error sending notifications", e);
+            }
+        }
+    }
+
+    private Map<String, String> getNotificationParametersMap() {
+        Map<String, String> paramMap = new HashMap<>();
+        if (StringUtils.isNotBlank(phabObjectId)) {
+            paramMap.put(PhabricatorParameters.ICON.getParamKey(), PhabricatorIcon.WARNING.toString());
```
[the remark this answers is not in the document]
Comment (maallen) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/PhabricatorExtractionCheckCommand.java:192]: Not for the extraction checks command but the notification service can be re-used by anyone so I wanted to provide an easy way to provide an icon that gets prefixed to a message

File: common/src/main/java/com/box/l10n/mojito/notifications/service/ThirdPartyNotificationService.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
+package com.box.l10n.mojito.notifications.service;
+
+import com.google.common.collect.ImmutableMap;
+import com.ibm.icu.text.MessageFormat;
+
+public abstract class ThirdPartyNotificationService {
+    
+    public abstract void sendMessage(String messageTemplate, ImmutableMap<String, String> serviceParameters) throws Exception;
```
[the remark this answers is not in the document]
Comment (maallen) [reply common/src/main/java/com/box/l10n/mojito/notifications/service/ThirdPartyNotificationService.java:8]: After looking at it again I agree, I'll clean this up so that a NotificationService.sendNotifications accepts just a message string & a set of services to notify with that message. If the messages are to differ across different service types then the calling object will need to handle that by calling the notification service with different parameters

File: common/src/main/java/com/box/l10n/mojito/notifications/service/NotificationService.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (maallen) [reply common/src/main/java/com/box/l10n/mojito/notifications/service/NotificationService.java:32]: We need the required = false piece as otherwise Spring will throw an exception on startup as there's no qualifying bean for the list

File: common/src/main/java/com/box/l10n/mojito/notifications/service/ThirdPartyNotificationService.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+    public abstract ThirdPartyNotification getThirdPartyNotificationServiceType();
```
[the remark this answers is not in the document]
Comment (maallen) [reply common/src/main/java/com/box/l10n/mojito/notifications/service/ThirdPartyNotificationService.java:10]: I've removed this method call in favour of adding the class type as a field on the enum and checking at runtime if we have a sender that is an instance of the class.

I don't believe that we can completely rely on spring configuration here as the runtime notification service needs to be able to differentiate between the available services for the use case that I don't want to notify all services with the same message e.g. I have both slack & phabricator enabled but I only want to notify phabricator with a specific message

**Review by aurambaj — COMMENTED** [review]
Put a bunch of comments most of them minor, the main point though is in the notification interface and the Phab/Slack implementation.  I put a small diff to illustrate [here](https://github.com/box/mojito/compare/notif_refacto?expand=1)

File: cli/src/main/java/com/box/l10n/mojito/cli/command/ExtractionCheckCommand.java [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+
+    @Parameter(names = {"--checks-skipped-message", "-csm"}, arity = 1, required = false, description = "Optional notification message to be sent when checks are skipped.")
+    String checksSkippedMessage;
+
+    @Override
+    protected void execute() throws CommandException {
+
+        if (areChecksSkipped) {
+            consoleWriter.fg(Ansi.Color.YELLOW).newLine().a("Checks disabled as --skip-checks is set to true.").println();
+            if (StringUtils.isNotBlank(checksSkippedMessage)) {
```
[the remark this answers is not in the document]
Comment (maallen) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/ExtractionCheckCommand.java:126]: isNotBlank returns false for a null parameter so I think we're fine for an NPE, I'll update to use Guava isNullOrEmpty and negate the check

File: common/src/main/java/com/box/l10n/mojito/notifications/service/ThirdPartyNotificationMessageSender.java [hunk 10]
```diff
@@ -0,0 +1,10 @@
+package com.box.l10n.mojito.notifications.service;
+
+import com.google.common.collect.ImmutableMap;
+import com.ibm.icu.text.MessageFormat;
+
+public abstract class ThirdPartyNotificationMessageSender {
+    
+    public abstract void sendMessage(String message, ImmutableMap<String, String> serviceParameters) throws Exception;
```
[the remark this answers is not in the document]
Comment (maallen) [reply common/src/main/java/com/box/l10n/mojito/notifications/service/ThirdPartyNotificationMessageSender.java:8]: I think the original intention of allowing an identical message to be sent to multiple service types here is complicating the logic too much, I like the idea of passing a message builder down with a receiver specified. If a caller wishes to send notifications to multiple services they can call the API repeatedly

File: cli/src/main/java/com/box/l10n/mojito/cli/command/ExtractionCheckCommand.java [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+    }
+
+    private void sendFailureNotifications(List<CliCheckResult> failures, boolean hardFail) {
+        ImmutableMap<String, String> notificationParams = getNotificationParametersMap("danger");
+
+        if (!thirdPartyNotificationTypes.isEmpty()) {
+            consoleWriter.newLine().a("Sending notifications").println();
+            for (ThirdPartyNotificationType notificationType : thirdPartyNotificationTypes) {
+                try {
+                    notificationService.sendNotifications(buildFailureNotificationMessage(failures, notificationType, hardFail), ImmutableSet.<ThirdPartyNotificationType>builder().add(notificationType).build(), notificationParams);
```
[the remark this answers is not in the document]
Comment (maallen) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/ExtractionCheckCommand.java:176]: Going to update the API to only send notifications to a single type. Multiple services can be sent notifications by calling the api with different parameters.

File: cli/src/main/java/com/box/l10n/mojito/cli/command/extractioncheck/ExtractionCheckNotificationSenderSlack.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+import org.springframework.beans.factory.annotation.Configurable;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Configurable
+public class ExtractionCheckNotificationSenderSlack extends ExtractionCheckNotificationSender {
+
+    @Autowired
+    ThirdPartyNotificationService thirdPartyNotificationService;
```
Comment (aurambaj) [inline cli/src/main/java/com/box/l10n/mojito/cli/command/extractioncheck/ExtractionCheckNotificationSenderSlack.java:21]: Wondering, what's the benefit of going through that service again?

it seems to be doing that:
```
       ---> slack                                 slack ---> 
CLI                    ---> generic --->
       ---> phab                                  phab --->
```

while it could just be?
```
       ---> slack
CLI      
       ---> phab

```

File: common/src/main/java/com/box/l10n/mojito/thirdpartynotification/ThirdPartyNotificationService.java [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.List;
+import java.util.stream.Stream;
+
+/**
+ * Notification service that sends messages to third party service providers.
+ *
+ * @author mallen
+ */
+@Component
+public class ThirdPartyNotificationService {
```
[the remark this answers is not in the document]
Comment (maallen) [reply common/src/main/java/com/box/l10n/mojito/thirdpartynotification/ThirdPartyNotificationService.java:18]: Moving the Notification service to common was to allow a common interface for notifications that could be reused on both server & client side in future, that was the reasoning behind introducing the NotificationService. Next time we need to send notifications from either cli/server it should be a case of injecting the notification service and writing a builder to construct your desired message format.

File: cli/pom.xml [hunk 14]
```diff
@@ -136,11 +136,33 @@
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-surefire-plugin</artifactId>
-                <configuration>
-                    <argLine>-Dspring.config.additional-location=file://${user.home}/.l10n/config/cli/
-                        -Dspring.profiles.active=${user.name},test -Xmx1024m
-                    </argLine>
-                </configuration>
+                <executions>
```
Comment (maallen) [inline cli/pom.xml:139]: I needed to introduce the separate executions here as with the previous config the tests hung when run as part of the maven build. I suspect it has something to do with the AspectJ @Configurable annotation not being setup in the test spring boot container as initially I was getting unsatisfied dependency errors when creating the new object, then when I used MockBeans and extended from CLITestBase the test hung on the RepoUpdateCommandTest so I put the notification tests in their own separate execution.

Comment (aurambaj) [reply cli/pom.xml:139]: I see. I don't think it is really maintainable to have profile just for a specific test. Not sure what you mean by "then when I used MockBeans and extended from CLITestBase". But iterating on what you were saying, 2 track to investigate:

1) Can change define a test configuration similar to what you did in S3FallbackImageServiceTest?

2) Or (but that would pollute the code), you can make @Autowired(required = false) in ExtractionCheckNotificationSenderSlack and just add them after the instance is created. The problem with that is the runtime code doesn't ensure the dependencies are present, hence some potential NPE that would need to be handled.

3) could refactor not to use @configurable, and just pass everything clearly via the sender constructor

 

I guess what was happening is that @Autowired by default is "required = true" and in that case you basically want to instantiate w

Comment (maallen) [reply cli/pom.xml:139]: Managed to get this back into a single execution profile by drawing inspiration from S3FallbackImageServiceTest so thanks for the suggestion! Issue is a container issue as suspected, there's no DifferentialRevision, SlackClient or SlackChannels beans in the container so AspectJ fails to inject them on creation of a new notifier object. 

I'm still unsure as to why @MockBean annotated on the fields didn't work on its own, but the fix was that I needed to annotate the test class with **@SpringBootTest** and add a static **@TestConfiguration** class that manually created the MockBeans in the container. I then needed to add the test class itself as a class loaded into the container using the **@SpringBootTest** annotation to ensure the TestConfiguration was picked and this allowed tests to pass in the maven build

**Review by aurambaj — APPROVED** [review]
